### PR TITLE
Feature/validate requirements script

### PR DESCRIPTION
**Pull request by mauro-cit**

### CODE REVIEW

File: 0-bootstrap/terraform.example.tfvars [hunk 1]
```diff
@@ -14,15 +14,19 @@
...
 
-group_billing_admins = "gcp-billing-admins@example.com"
+group_billing_admins = "REPLACE_ME"
 
-default_region = "us-central1"
+# Example filling
+# group_org_admins = "gcp-organization-admins@example.com"
+# group_billing_admins = "gcp-billing-admins@example.com"
+
+default_region = "REPLACE_ME"
```
Comment (daniel-cit) [inline 0-bootstrap/terraform.example.tfvars]: "us-central1" ia a valid default value, it does not not need to be `REPLACED`

File: 0-bootstrap/terraform.example.tfvars [hunk 2]
```diff
@@ -14,15 +14,19 @@
...
+billing_account = "REPLACE_ME"
 
-group_org_admins = "gcp-organization-admins@example.com"
+group_org_admins = "REPLACE_ME"
 
-group_billing_admins = "gcp-billing-admins@example.com"
+group_billing_admins = "REPLACE_ME"
 
-default_region = "us-central1"
+# Example filling
```
Comment (daniel-cit) [inline 0-bootstrap/terraform.example.tfvars]: I
```suggestion
# Example of values for the groups
```

File: scripts/validate-requirements.sh [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+GIT_VERSION="2.25.1"
+# User Inputs
+END_USER_CREDENTIAL="$1"
+ORGANIZATION_ID="$2"
+BILLING_ACCOUNT="$3"
+
+ERRORS=""
+# -------------------------- Funcions ---------------------------
+
+# Compare the two float numbers
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]:
```suggestion
# Compare two semantic versions
```

File: scripts/validate-requirements.sh [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+    if [ ! "$(command -v terraform )" ]; then
+        echo "Terraform not found."
+        echo "Visit https://learn.hashicorp.com/tutorials/terraform/install-cli and follow the instructions to install Terraform."
+        ERRORS+=$'Terraform not found\n'
+    else
+        TERRAFORM_CURRENT_VERSION=$(terraform version -json | jq -r .terraform_version)
+        compare_version "$TERRAFORM_CURRENT_VERSION" "$TF_VERSION"
+        if [ $? -eq 2 ]; then
+            echo "An error was found with the Terraform installation."
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]:
```suggestion
            echo "An incompatibly was found in Terraform version."
```

File: scripts/validate-requirements.sh [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            echo "Visit https://learn.hashicorp.com/tutorials/terraform/install-cli and follow the instructions to install Terraform."
+            ERRORS+=$'Terraform version is incompatible.\n'
+        fi
+    fi
+}
+
+# Validate the Google Cloud SDK installation and version
+function validate_gcloud(){
+    if [ ! "$(command -v gcloud)" ]; then
+        echo "Gcloud not found."
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]:
```suggestion
        echo "gcloud CLI not found."
```

File: scripts/validate-requirements.sh [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        echo "Visit https://cloud.google.com/sdk/docs/install and follow the instructions to install gcloud CLI."
+        ERRORS+=$'gcloud not found.\n'
+    else
+        GCLOUD_CURRENT_VERSION=$(gcloud version --format=json | jq -r '."Google Cloud SDK"')
+        compare_version "$GCLOUD_CURRENT_VERSION" "$GCLOUD_SDK_VERSION"
+        if [ $? -eq 2 ]; then
+            echo "Gcloud version $GCLOUD_SDK_VERSION is required."
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]:
```suggestion
            echo "gcloud version $GCLOUD_SDK_VERSION is required."
```

File: scripts/validate-requirements.sh [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+function validate_git(){
+    if [ ! "$(command -v git)" ]; then
+        echo "git not found."
+        echo "Visit https://git-scm.com/book/en/v2/Getting-Started-Installing-Git and follow the instructions to install Git."
+        ERRORS+=$'git not found.\n'
+    else
+        GIT_CURRENT_VERSION=$(git version | awk '{print $3}')
+        compare_version "$GIT_CURRENT_VERSION" "$GIT_VERSION"
+        if [ $? -eq 2 ]; then
+            echo "An error was found with the git installation."
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]:
```suggestion
            echo "An incompatibly was found in git version."
```

File: scripts/validate-requirements.sh [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            echo "Version required is $GIT_VERSION"
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]:
```suggestion
            echo "Version required is al least $GIT_VERSION"
```

File: scripts/validate-requirements.sh [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+# Validate the Terraform installation and version
+function validate_terraform(){
+
+    if [ ! "$(command -v terraform )" ]; then
+        echo "Terraform not found."
+        echo "Visit https://learn.hashicorp.com/tutorials/terraform/install-cli and follow the instructions to install Terraform."
+        ERRORS+=$'Terraform not found\n'
+    else
+        TERRAFORM_CURRENT_VERSION=$(terraform version -json | jq -r .terraform_version)
+        compare_version "$TERRAFORM_CURRENT_VERSION" "$TF_VERSION"
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: Terraform version comparation is different from the others:
- `terraform` version must be `==` 
- `gcloud` and `git` must be `>=`

File: scripts/validate-requirements.sh [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    check_billing_account_roles "$END_USER_CREDENTIAL" "$BILLING_ACCOUNT"
+}
+
+# Verifies wheter a user is assigned to the expected Orgzanization Level roles
+function check_org_level_roles(){
+
+    gcloud organizations get-iam-policy "$2" \
+    --filter="bindings.members:$1" \
+    --flatten="bindings[].members" \
+    --format="table(bindings.role)" 2>/dev/null > org-level-roles-output.txt
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: is it possible to save this data in memory?

File: scripts/validate-requirements.sh [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+    lines=$(grep -c -e roles/resourcemanager.folderCreator -e roles/resourcemanager.organizationAdmin org-level-roles-output.txt)
+    if [ "$lines" -ne 2 ]; then
+        echo "The User must have the Organization Roles resourcemanager.folderCreator and resourcemanager.organizationAdmin"
+        ERRORS+=$'There are missing organization level roles on the Credential.\n'
+    fi
+    rm -Rf org-level-roles-output.txt
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: can we do it in memory

File: scripts/validate-requirements.sh [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+# Verifies wheter a user is assigned to the expected Billing Level roles
+function check_billing_account_roles(){
+
+    gcloud beta billing accounts get-iam-policy "$2" \
+        --filter="bindings.members:$1" \
+        --flatten="bindings[].members" \
+        --format="table(bindings.role)" 2>/dev/null > billing-account-roles.txt
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: can we do it in memory?

File: scripts/validate-requirements.sh [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+    lines=$(grep -c roles/billing.admin billing-account-roles.txt)
+    if [ "$lines" -ne 1 ]; then
+        echo "The User must have the Billing Account Role billing.admin"
+        ERRORS+=$'There are missing billing account level roles on the Credential.\n'
+    fi
+    rm -Rf billing-account-roles.txt
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: can we do it in memory?

File: scripts/validate-requirements.sh [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+echo "......................................."
+if [ -z "$ERRORS" ]; then
+    echo "Validation successfull!"
+    echo "No errors found."
+else
+    echo "Validation failed!"
+    echo "Errors found:"
+    echo "$ERRORS"
+fi
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: this could be in a main function

File: scripts/validate-requirements.sh [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+# -------------------------- Variables --------------------------
+# Expected versions of the installers
+TF_VERSION="0.13.7"
+GCLOUD_SDK_VERSION="319.0.0"
+GIT_VERSION="2.25.1"
+# User Inputs
+END_USER_CREDENTIAL="$1"
+ORGANIZATION_ID="$2"
+BILLING_ACCOUNT="$3"
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: you clous check that that these values are not null before running
and present a helper message

File: scripts/validate-requirements.sh (hunk 10 above)
Comment (mauro-cit) [reply scripts/validate-requirements.sh]: Yes, just found a way for doing so. Now I'll replicate to the other cases.

File: 0-bootstrap/terraform.example.tfvars [hunk 16]
```diff
@@ -14,13 +14,17 @@
  * limitations under the License.
  */
 
-org_id = "000000000000"
+org_id = "REPLACE_ME"
 
-billing_account = "000000-000000-000000"
+billing_account = "REPLACE_ME"
```
Comment (daniel-cit) [inline 0-bootstrap/terraform.example.tfvars]:
```suggestion
billing_account = "REPLACE_ME" # format "000000-000000-000000"
```

File: 0-bootstrap/terraform.example.tfvars [hunk 17]
```diff
@@ -14,13 +14,17 @@
  * limitations under the License.
  */
 
-org_id = "000000000000"
+org_id = "REPLACE_ME"
```
Comment (daniel-cit) [inline 0-bootstrap/terraform.example.tfvars]:
```suggestion
org_id = "REPLACE_ME" # format "000000000000"
```

File: scripts/validate-requirements.sh [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+        if [ "$(grep -c REPLACE_ME $FILE)" != 0 ]; then
+            echo "$FILE must have required values fullfiled."
+            ERRORS+=$'terraform.tfvars file must be correctly fullfiled for 0-bootstrap step.\n'
+        fi
+    fi
+}
+
+function main(){
+
+    if [ -z "$END_USER_CREDENTIAL" ] || [ -z "$ORGANIZATION_ID" ] || [ -z "$BILLING_ACCOUNT" ]; then
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: it looks better if you create and use a `usage` function like in
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/616ede9456cc8f86ef7995192af3473d17ee7946/helpers/setup-sa.sh#L26

```
usage() {
  echo
  echo " Usage:"
  echo "     $0 -o <organization id> -p <project id> [-b <billing account id>] [-f <folder id>] [-n <service account name>]"
  echo "         organization id        (required)"
  echo "         project id             (required)"
  echo "         billing account id     (optional)"
  echo "         folder id              (optional)"
  echo "         service account name   (optional)"
  echo
  echo " The billing account id is required if owned by a different organization"
  echo " than the Seed Project organization."
  echo
  exit 1
}
```

File: scripts/validate-requirements.sh [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+# Usage:
+# bash scripts/validate-requirements.sh "END_USER_EMAIL" "ORGANIZATION_ID" "BILLING_ACCOUNT_ID"
+
+# -------------------------- Variables --------------------------
+# Expected versions of the installers
+TF_VERSION="1.2.1"
+GCLOUD_SDK_VERSION="319.0.0"
+GIT_VERSION="2.25.1"
+# User Inputs
+END_USER_CREDENTIAL="$1"
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: you should use `getopts` to read the input from the user

https://github.com/terraform-google-modules/terraform-google-data-fusion/blob/52d8178c29cc2491d799e63ae1bea5a688b29773/modules/wait_healthy/wait_healthy.sh#L21

```
while getopts 'e:t:' c
do
  case $c in
    e) ENDPOINT=${OPTARG} ;;
    t) ACCESS_TOKEN=${OPTARG} ;;
    *)
      echo "Invalid flag ${OPTARG}"
      exit 1
      ;;
  esac
done
```

File: scripts/validate-requirements.sh [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Usage:
+# bash scripts/validate-requirements.sh "END_USER_EMAIL" "ORGANIZATION_ID" "BILLING_ACCOUNT_ID"
+
+# -------------------------- Variables --------------------------
+# Expected versions of the installers
+TF_VERSION="1.2.1"
```
Comment (daniel-cit) [inline scripts/validate-requirements.sh]: required version is

[Terraform](https://www.terraform.io/downloads.html) version 0.13.7.

File: scripts/validate-requirements.sh (hunk 18 above)
Comment (mauro-cit) [reply scripts/validate-requirements.sh]: Awesome idea, I haven't think about something like this

**Pull request closed without merging**